src/librbd/image/RemoveRequest.cc

```diff
@@ -0,0 +1,764 @@
+                 << dendl;
+    send_close_image(*result);
+    return nullptr;
+  }
+
+  if (!m_old_format) {
+    send_disable_mirror();
+    return nullptr;
+  }
+  send_close_image(*result);
+  return nullptr;
+}
+
+template<typename I>
+void RemoveRequest<I>::send_disable_mirror() {
+  ldout(m_cct, 20) << dendl;
+
+  using klass = RemoveRequest<I>;
+  Context *ctx = create_context_callback<klass, &klass::handle_disable_mirror>(this);
+
+  mirror::DisableRequest<I> *req =
+    mirror::DisableRequest<I>::create(m_image_ctx, m_force, !m_force, ctx);
+  req->send();
+}
+
+template<typename I>
+Context *RemoveRequest<I>::handle_disable_mirror(int *result) {
+  ldout(m_cct, 20) << ": r=" << *result << dendl;
+
+
+  if (*result == -EOPNOTSUPP) {
+    *result = 0;
+  } else if (*result < 0) {
+    lderr(m_cct) << "error disabling image  mirroring: " << cpp_strerror(*result)
+                 << dendl;
+  }
+
+  send_close_image(*result);
+  return nullptr;
+}
+
+template<typename I>
+void RemoveRequest<I>::send_close_image(int r) {
+  ldout(m_cct, 20) << dendl;
+
+  m_retval = r;
+  assert(!m_image_ctx->owner_lock.is_locked());
+
+  using klass = RemoveRequest<I>;
+  Context *ctx = create_context_callback<klass, &klass::handle_send_close_image>(this);
+
+  m_image_ctx->state->close(ctx);
+}
+
+template<typename I>
+Context *RemoveRequest<I>::handle_send_close_image(int *result) {
+  ldout(m_cct, 20) << ": r=" << *result << dendl;
+
+  if (*result < 0) {
+    lderr(m_cct) << "error encountered while closing image: " << cpp_strerror(*result)
+                 << dendl;
+  }
+
+  if (m_retval < 0) {
+    switch_thread_context();
+  } else {
+    remove_header();
+  }
+  return nullptr;
+}
+
+template<typename I>
+void RemoveRequest<I>::remove_header() {
+  ldout(m_cct, 20) << dendl;
+
+  using klass = RemoveRequest<I>;
+  librados::AioCompletion *rados_completion =
+    create_rados_ack_callback<klass, &klass::handle_remove_header>(this);
+  int r = m_ioctx.aio_remove(m_header_oid, rados_completion);
+  assert(r == 0);
+  rados_completion->release();
+}
+
+template<typename I>
+Context *RemoveRequest<I>::handle_remove_header(int *result) {
+  ldout(m_cct, 20) << ": r=" << *result << dendl;
+
+  if ((*result < 0) && (*result != -ENOENT)) {
+    lderr(m_cct) << "error removing header: " << cpp_strerror(*result) << dendl;
+  }
+
+  switch_thread_context();
+  return nullptr;
+}
+
+template <typename I>
+void RemoveRequest<I>::switch_thread_context() {
+  ldout(m_cct, 20) << dendl;
+
+  using klass = RemoveRequest<I>;
+
+  Context *ctx = create_context_callback<klass, &klass::handle_switch_thread_context>(this);
+  m_op_work_queue->queue(ctx, 0);
+}
+
+template <typename I>
+void RemoveRequest<I>::handle_switch_thread_context(int r) {
+  ldout(m_cct, 20) << ": r=" << r << dendl;
+
+  m_image_ctx->destroy();
+  m_image_ctx = nullptr;
+
+  if (m_retval < 0 && m_retval != ENOENT && !m_unknown_format) {
```

Comment:
Not sure about that ```!m_unknown_format``` conditional -- if you fail to open the image due to -EPERM, you should fail out here but since ```m_unknown_format``` would be true in that case, you will continue to remove the image.